source/common/http/utility.cc

```diff
@@ -463,13 +466,22 @@ void Utility::sendLocalReply(const bool& is_reset, const EncodeFunctions& encode
       }
       response_headers->setGrpcMessage(PercentEncoding::encode(body_text));
     }
+    // The `modify_headers` function may have added content-length, remove it.
+    response_headers->removeContentLength();
     encode_functions.encode_headers_(std::move(response_headers), true); // Trailers only response
     return;
   }
 
   if (!body_text.empty()) {
     response_headers->setContentLength(body_text.size());
-    response_headers->setReferenceContentType(content_type);
+    // If the `rewrite` function has changed body_text or content-type is not set, set it.
+    if (body_text != local_reply_data.body_text_ || response_headers->ContentType() == nullptr) {
+      response_headers->setReferenceContentType(content_type);
+    }
+  } else {
```

Comment:
Let's make sure all of this new logic is unit tested.

Reply:
Updated code is not adding too much new logic.  The logic is covered by router.direct_response and ext_authz integration tests.